### DISCUSSION TITLE
Degradation models for lithium batteries

### DISCUSSION THREAD

**Discussion by wigging** [pybamm-team/PyBaMM#1597] [Ideas]
Does PyBaMM have aging/degradation models for lithium batteries? I don't see anything in the documentation about this capability. A paper by Reniers, Mulder, and Howey (see reference below) gives an overview of various degradation models and provides a C++ code called SLIDE (see GitHub link below). Since I prefer Python more than C++, I'm hoping that PyBaMM would have similar features as SLIDE.

1. Jorn M. Reniers, Grietus Mulder, and David A. Howey. "Review and Performance Comparison of Mechanical-Chemical Degradation Models for Lithium-Ion Batteries." Journal of The Electrochemical Society, vol. 166, no. 14, pp. A3189-A3200, 2019. https://doi.org/10.1149/2.0281914jes
2. SLIDE. A C++ code that simulates degradation of lithium ion cells. https://github.com/davidhowey/SLIDE

**Comment by valentinsulzer** [comment]
You can do degradation modeling by changing the options you pass to the model https://pybamm.readthedocs.io/en/latest/source/models/base_models/base_battery_model.html#pybamm.BatteryModelOptions

See the example notebooks for some examples. If there is a specific degradation model you are after and it isn't there you can open a new issue

**Reply by wigging** [reply]
Thank you for the comment. I'll try the example notebooks and let you know if I still have questions.

**Reply by alihassan90** [reply]
@tinosulzer i cant access this link. can you repaste?

**Reply by brosaplanella** [reply]
This is the updated link: https://pybamm.readthedocs.io/en/latest/source/api/models/base_models/base_battery_model.html#pybamm.BatteryModelOptions

**Comment by NLithium** [comment]
I will continue on this topic. I want to set up own submodel for calculating SEI formation and SEI reformation, but I have troubles correctly implementing it into PyBaMM. For example, how to correctly implement this submodel in PyBaMM - for example the SEI formation:
![image](https://user-images.githubusercontent.com/95873843/159665479-830fc3cb-03c6-4171-a827-f5d2fcfff818.png)

where resistance is defined as
![image](https://user-images.githubusercontent.com/95873843/159665550-eca5a88a-c185-4fc5-940b-45ea4cad2a4a.png)
![image](https://user-images.githubusercontent.com/95873843/159665636-4ae6d59e-b5fe-4b63-9b6d-60e30beb9335.png)

and exchange current density is defined as:
![image](https://user-images.githubusercontent.com/95873843/159665992-2f1a93d7-b7a6-4a81-8bfc-4c96890b639e.png)

equations are based on Journal of The Electrochemical Society, 164 (12) E287-E294 (2017).

**Reply by valentinsulzer** [reply]
If you're happy to edit the source code, you need to add a new SEI option [here](https://github.com/pybamm-team/PyBaMM/blob/23157aebce218444edc83b525dfb2c7fc8637598/pybamm/models/full_battery_models/base_battery_model.py#L237) and then a new if branch [here](https://github.com/pybamm-team/PyBaMM/blob/23157aebce218444edc83b525dfb2c7fc8637598/pybamm/models/submodels/interface/sei/sei_growth.py#L84). There is also a way to do this without editing source code, let me know if that's what you're looking to do

**Reply by valentinsulzer** [reply]
You'll also need to non-dimensionalize your SEI model

**Reply by valentinsulzer** [reply]
This looks quite similar to the reaction-limited model we already have actually. What are the equations for Q_crack and Q_Li2CO3?

**Reply by NLithium** [reply]
I have manage to implement it and currently running test cases. Looking similar to the reaction limited but the result is different.
Q_crack & Q_Li2CO3 stand for charge quantities.

Do you also have documentation which publications you have used for e.g. reaction-limited and similar submodels?

**Reply by NLithium** [reply]
@tinosulzer you have also said there is a way to change it without editing a source code. How can I do that?
Thank you!

**Reply by valentinsulzer** [reply]
Chapter 5 for SEI refs:
[Marquis - 2020 - Long-Term Degradation of Lithium-ion Batteries.pdf](https://github.com/pybamm-team/PyBaMM/files/8362761/Marquis.-.2020.-.Long-Term.Degradation.of.Lithium-ion.Batteries.pdf)

**Reply by valentinsulzer** [reply]
To change without editing source code you need to write your own `SEI` class that implements all the same methods as the sei growth class in pybamm, then do
```python
model = pybamm.lithium_ion.DFN(build=false)
model.submodels["sei"] = SEI(model.param)
model.build_model()
```
See https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/models/using-submodels.ipynb

**Reply by NLithium** [reply]
@tinosulzer Thank you very much for your answer!
Currently I am still working with source code, because we are looking for implementation of specific equations.

**Reply by Brushnie33** [reply]
@tinosulzer Hello, I was trying to execute the EC-reaction limited SEI model by writing it as a class, and then use the following lines:
model = pybamm.lithium_ion.DFN(build=False)
model.submodels["Negative sei"] = SEI_new_option(model.param, "negative","full electrode", model.options)
model.build_model()

I tried to compare the standalone class and the EC-reaction limited SEI model in the existing SEI growth model, the SEI capacity loss and thickness match with each other, however, I couldn't get the total lithium lost correct: 

![image](https://github.com/pybamm-team/PyBaMM/assets/112187928/2c662428-fd77-4539-ab4a-1789e3dbfbf3)

Do you have any suggestions on how to modify the codes? 
I copied class BaseInterface(pybamm.BaseSubModel) and BaseModel(BaseInterface) from the source codes, and then define SEI_new_option class for the EC-reaction limited model, do I need to add SEI_new_option in the "SEI" section in base_battery_model.py source codes?

**Reply by Brushnie33** [reply]
I added the SEI_new_option in the "SEI" section in base_battery_model.py and now it works properly.

**Comment by NLithium** [comment]
Additionally, I have a follow up questions.

1.  I'm attempting to implement the "not equal" operator '>' in the parameter function. The idea is to adjust the plating/stripping equation according on the potential sign.
![image](https://user-images.githubusercontent.com/95873843/160604955-b6ac8d2c-be1d-430f-95c7-12ea80afca7b.png)

Phi is variables["Negative electrode surface potential difference"]
How can I get the actual value of phi(number) from symbols in a parameter function?
Or there is another way to adjust some equation based on the potential sign with boundary condition?

2. I wanted to run a case with experiment using ```Chen2020``` & ```Mohtat2020``` parameters with ```okane2020_Li_plating``` parameters
```math
experiment = pybamm.Experiment(
            ["Charge at C/2 until 4.2 V",
            "Hold at 4.2 V until C/10",
            "Rest for 5 minutes",
            "Discharge at C/2 until 2.8 V",
            "Rest for 5 minutes"] * 10
```

```math
model = pybamm.lithium_ion.DFN(
    {   
        "cell geometry": "arbitrary",
        "SEI": "reaction limited",
        "SEI film resistance": "average",
        "SEI porosity change": "true",
        "lithium plating": "reversible",
        "lithium plating porosity change": "true"
        "thermal": "lumped"
    }
)
```

The output I get for ```Chen2020``:
"psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed"

The output I get for ```Mohtat2020``:
"Experiment is infeasible: 'event: Minimum voltage' was triggered during 'Charge at 1 C until 4.2 V'. The returned solution only contains the first 0 cycles. Try reducing the current, shortening the time interval, or reducing the period."

How can resolve these issues?

**Comment by xcblovecy** [comment]
You attemped to implement the "not equal" operator '>' in the parameter function. The idea is to adjust the plating/stripping equation according on the potential sign.  Can this approach be implemented?
